File: web-app/server/services/barcodeService.js
/**
 * Barcode Service Module
 * Handles barcode and QR code generation
 */

const logger = require('../logger');

/**
 * Generate a barcode value for a product
 * @param {string} productName - Product name
 * @param {string} productId - Product ID
 * @returns {string} Generated barcode value
 */
function generateProductBarcode(productName, productId) {
  // Generate a simple barcode based on product name and ID
  const prefix = 'PROD';
  const idPart = productId.substring(productId.length - 8); // Last 8 chars of ID
  const namePart = productName.substring(0, 3).toUpperCase().replace(/[^A-Z0-9]/g, '');
  return `${prefix}${namePart}${idPart}`.substring(0, 20);
}

/**
 * Generate barcode image (base64 encoded)
 * Note: This is a placeholder. For production, use a library like 'bwip-js' or 'jsbarcode'
 * @param {string} barcodeValue - The barcode value to encode
 * @returns {Promise<string>} Base64 encoded barcode image
 */
async function generateBarcodeImage(barcodeValue) {
  try {
    // Placeholder implementation
    // In production, integrate with a barcode library like:
    // const bwipjs = require('bwip-js');
    // const png = await bwipjs.toBuffer({
    //   bcid: 'code128',
    //   text: barcodeValue,
    //   scale: 3,
    //   height: 10,
    //   includetext: true
    // });
    // return `data:image/png;base64,${png.toString('base64')}`;
    
    logger.warn('Barcode generation not implemented - returning placeholder');
    return `data:image/svg+xml;base64,${Buffer.from(
      `<svg width="200" height="80" xmlns="http://www.w3.org/2000/svg">
        <text x="10" y="40" font-family="monospace" font-size="14">${barcodeValue}</text>
      </svg>`
    ).toString('base64')}`;
  } catch (error) {
    logger.error('Barcode generation error:', error);
    throw new Error('Failed to generate barcode');
  }
}

/**
 * Generate QR code image (base64 encoded)
 * Note: This is a placeholder. For production, use a library like 'qrcode'
 * @param {Object} data - Data to encode in QR code
 * @returns {Promise<string>} Base64 encoded QR code image
 */
async function generateQRCode(data) {
  try {
    // Placeholder implementation
    // In production, integrate with QR code library:
    // const QRCode = require('qrcode');
    // const qrCode = await QRCode.toDataURL(JSON.stringify(data));
    // return qrCode;
    
    logger.warn('QR code generation not implemented - returning placeholder');
    const dataStr = JSON.stringify(data);
    return `data:image/svg+xml;base64,${Buffer.from(
      `<svg width="200" height="200" xmlns="http://www.w3.org/2000/svg">
        <rect width="200" height="200" fill="white"/>
        <text x="10" y="100" font-family="monospace" font-size="10">${dataStr.substring(0, 30)}</text>
      </svg>`
    ).toString('base64')}`;
  } catch (error) {
    logger.error('QR code generation error:', error);
    throw new Error('Failed to generate QR code');
  }
}

module.exports = {
  generateProductBarcode,
  generateBarcodeImage,
  generateQRCode
};

File: server-flask/services/barcode_service.py
import logging
import base64

logger = logging.getLogger(__name__)

def generate_product_barcode(product_name, product_id):
    """
    Generate a simple textual barcode identifier based on product name and ID
    """
    prefix = 'PROD'
    id_part = str(product_id)[-8:] if len(str(product_id)) >= 8 else str(product_id)
    name_part = ''.join(e for e in str(product_name)[:3].upper() if e.isalnum())
    
    result = f"{prefix}{name_part}{id_part}"
    return result[:20]

def generate_barcode_image(barcode_value):
    """
    Generate a base64 encoded SVG barcode placeholder (matching legacy JS behavior)
    """
    try:
        logger.warning('Barcode generation not fully implemented - returning SVG placeholder')
        svg_content = f"""<svg width="200" height="80" xmlns="http://www.w3.org/2000/svg">
            <text x="10" y="40" font-family="monospace" font-size="14">{barcode_value}</text>
        </svg>"""
        
        b64_encoded = base64.b64encode(svg_content.encode('utf-8')).decode('utf-8')
        return f"data:image/svg+xml;base64,{b64_encoded}"
    except Exception as e:
        logger.error(f'Barcode generation error: {e}')
        raise Exception('Failed to generate barcode')

def generate_qr_code(data):
    """
    Generate a base64 encoded SVG QR placeholder (matching legacy JS behavior)
    """
    try:
        logger.warning('QR code generation not fully implemented - returning SVG placeholder')
        
        # Simple string representation of dict
        import json
        data_str = json.dumps(data)[:30]
        
        svg_content = f"""<svg width="200" height="200" xmlns="http://www.w3.org/2000/svg">
            <rect width="200" height="200" fill="white"/>
            <text x="10" y="100" font-family="monospace" font-size="10">{data_str}</text>
        </svg>"""
        
        b64_encoded = base64.b64encode(svg_content.encode('utf-8')).decode('utf-8')
        return f"data:image/svg+xml;base64,{b64_encoded}"
    except Exception as e:
        logger.error(f'QR code generation error: {e}')
        raise Exception('Failed to generate QR code')
